### PR TITLE
Pytorch bindings

### PR DESCRIPTION
**Pull request by SeanNaren**
Hopefully looks good, I'm having difficulties getting the GPU implementation to work so any feedback please let me know!

### CODE REVIEW

**Comment by SeanNaren** [comment]
Should be good to go! Let me know of any other changes I need to make. Slight off topic, we should really get the pascal flags merged into the main repo, or atleast:

`set(CUDA_NVCC_FLAGS "${CUDA_NVCC_FLAGS} -gencode arch=compute_60,code=sm_60")` added to the CmakeList!

**Comment by SeanNaren** [comment]
@jaredcasper any chance I could get a review on this PR?

**Comment by jaredcasper** [comment]
@SeanNaren Sorry, I recently left Baidu so no longer have permissions for this repo.

**Comment by stepelu** [comment]
It doesn't build on clang and OSX latest (10.12.3):

```
> python setup.py install
CUDA_HOME not found in the environment so building without GPU support. To build with GPU support please define the CUDA_HOME environment variable. This should be a path which contains include/cuda.h
generating build/_warp_ctc.c
(already up-to-date)
not modified: 'build/_warp_ctc.c'
running install
running build
running build_py
creating build/lib.macosx-10.11-x86_64-3.5
creating build/lib.macosx-10.11-x86_64-3.5/warpctc_pytorch
copying warpctc_pytorch/__init__.py -> build/lib.macosx-10.11-x86_64-3.5/warpctc_pytorch
running build_ext
building 'warpctc_pytorch._warp_ctc' extension
creating build/temp.macosx-10.11-x86_64-3.5
creating build/temp.macosx-10.11-x86_64-3.5/build
creating build/temp.macosx-10.11-x86_64-3.5/Users
creating build/temp.macosx-10.11-x86_64-3.5/Users/sp
creating build/temp.macosx-10.11-x86_64-3.5/Users/sp/Prj
creating build/temp.macosx-10.11-x86_64-3.5/Users/sp/Prj/warp-ctc
creating build/temp.macosx-10.11-x86_64-3.5/Users/sp/Prj/warp-ctc/pytorch_binding
creating build/temp.macosx-10.11-x86_64-3.5/Users/sp/Prj/warp-ctc/pytorch_binding/src
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/sp/env/lib/python3.5/site-packages/torch/utils/ffi/../../lib/include -I/Users/sp/env/lib/python3.5/site-packages/torch/utils/ffi/../../lib/include/TH -I/Users/sp/Prj/warp-ctc/include -I/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/include/python3.5m -c build/_warp_ctc.c -o build/temp.macosx-10.11-x86_64-3.5/build/_warp_ctc.o -std=c++11 -fPIC
error: invalid argument '-std=c++11' not allowed with 'C/ObjC'
error: command 'clang' failed with exit status 1
```

Clang info:
```
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin16.4.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

**Comment by stepelu** [comment]
Also, it seems that the binding should check that the activations are contiguous (a wrong cost is returned otherwise but no error it thrown).

**Comment by SeanNaren** [comment]
@stepelu agreed, will try address these issues!

**Comment by geffy** [comment]
Will it be reviewed/merged? 
Seems like very useful feature.

**Comment by SeanNaren** [comment]
@geffy quite honestly I'm not sure who I can nag anymore to pull this in...

**Comment by geffy** [comment]
@gangliao @dxyzab can you please review this or suggest someone who will?

**Comment by ryanleary** [comment]
Bump.

**Comment by stepelu** [comment]
Bump

**Comment by igoror** [comment]
@SeanNaren inside pytorch_binding/setup.py value of CUDA_HOME should be used to add additional include directories for cuda.

File: pytorch_binding/src/binding.cpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+                     labels_ptr, label_sizes_ptr,
+                     sizes_ptr, probs->size[2],
+                     minibatch_size, costs_ptr,
+                     cpu_workspace, options);
+
+    delete cpu_workspace;
+    return 1;
+}
+#ifdef WARPCTC_ENABLE_GPU
+   extern "C" int gpu_ctc(THCudaTensor *probs,
```
Comment (howard0su) [inline pytorch_binding/src/binding.cpp]: too much duplicate code with cpu_ctc, can u make a common function and let cpu and gpu functions call the common function to avoid duplication?

File: pytorch_binding/src/binding.cpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    // have to use at least one
+    options.num_threads = std::max(options.num_threads, (unsigned int) 1);
+#endif
+
+    size_t cpu_size_bytes;
+    get_workspace_size(label_sizes_ptr, sizes_ptr,
+                       (int) probs->size[2], minibatch_size,
+                       options, &cpu_size_bytes);
+
+    float* cpu_workspace = (float*) new unsigned char[cpu_size_bytes];
```
Comment (howard0su) [inline pytorch_binding/src/binding.cpp]: float* cpu_workspace = new float[cpu_size_bytes/sizeof(float)];

File: pytorch_binding/README.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+git clone https://github.com/baidu-research/warp-ctc.git
+cd warp-ctc
+mkdir build; cd build
+cmake ..
+make
+```
+
+Otherwise, set `WARP_CTC_PATH` to wherever you have `libwarpctc.so`
+installed. If you have a GPU, you should also make sure that
+`CUDA_HOME` is set to the home cuda directory (i.e. where
```
Comment (howard0su) [inline pytorch_binding/README.md]: the home directory of cuda

File: pytorch_binding/README.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+`include/cuda.h` and `lib/libcudart.so` live). For example:
+
+```
+export CUDA_HOME="/usr/local/cuda"
+```
+
+Now install the bindings:
```
Comment (howard0su) [inline pytorch_binding/README.md]: binding, no 's'

File: pytorch_binding/warpctc_pytorch/__init__.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+
+class _CTC(Function):
+    def forward(self, acts, labels, act_lens, label_lens):
+        is_cuda = True if acts.is_cuda else False
+        acts = acts.contiguous()
+        loss_func = warp_ctc.gpu_ctc if is_cuda else warp_ctc.cpu_ctc
+        grads = torch.zeros(acts.size()).type_as(acts)
+        minibatch_size = acts.size(1)
+        costs = torch.zeros(minibatch_size).cpu()
```
Comment (howard0su) [inline pytorch_binding/warpctc_pytorch/__init__.py]: why call cpu() here? zeros should return tensor on CPU anyway.

**Comment by huang-aoyun** [comment]
I set the CUDA_HOME environment,but there  is still have the problem as follows,can you help me?
CUDA_HOME not found in the environment so building without GPU support. To build with GPU support please define the CUDA_HOME environment variable. This should be a path which contains include/cuda.h
generating build/_warp_ctc.c
(already up-to-date)
not modified: 'build/_warp_ctc.c'
running install
running build
running build_py
running build_ext
running install_lib
running install_egg_info
Removing /usr/local/lib/python2.7/dist-packages/warpctc_pytorch-0.1.egg-info
Writing /usr/local/lib/python2.7/dist-packages/warpctc_pytorch-0.1.egg-info
